language/testing-infra/functional-tests/src/testsuite.rs

```diff
@@ -79,167 +93,227 @@ pub fn functional_tests<TComp: Compiler>(
-                at_most_n_before_and_m_after(&res.text[m.entry_id], 30, m.start, m.end, 50);
-            write!(output, "{}", indent)?;
-            write!(output, "{}{}", prefix, before)?;
-            output.set_color(ColorSpec::new().set_underline(true).set_fg(Some(
-                if $is_positive {
-                    Color::Green
-                } else {
-                    Color::Red
-                },
-            )))?;
-            write!(output, "{}", matched)?;
-            output.reset()?;
-            writeln!(output, "{}", after)?;
+    let term_width = match term_size::dimensions() {
+        Some((w, _h)) => w,
+        _ => 80,
+    };
 
-            let offset = prefix.chars().count() + before.chars().count();
-            write!(output, "{}", indent)?;
-            write!(
+    macro_rules! print_horizontal_line {
+        () => {
+            writeln!(
                 output,
                 "{}",
-                iter::repeat(' ').take(offset).collect::<String>()
-            )?;
-            print_directive!(m.pat_id)?;
+                iter::repeat('=').take(term_width).collect::<String>()
+            )
+        };
+    }
+
+    macro_rules! print_test_header {
+        () => {{
+            writeln!(output)?;
+            print_horizontal_line!()?;
+            writeln!(output, "{}", path.display())?;
             writeln!(output)
         }};
     }
 
-    writeln!(output)?;
-    writeln!(
-        output,
-        "{}",
-        iter::repeat('=').take(100).collect::<String>()
-    )?;
-    writeln!(output, "{}", path.display())?;
-    writeln!(output)?;
-
-    // Render the evaluation log.
-    output.set_color(ColorSpec::new().set_bold(true).set_fg(Some(Color::Yellow)))?;
-    write!(output, "info: ")?;
-    output.set_color(ColorSpec::new().set_bold(true))?;
-    writeln!(output, "Evaluation Outputs")?;
-    output.reset()?;
-    if pretty_mode() {
-        writeln!(
-            output,
-            "{}",
-            log.outputs
-                .iter()
-                .enumerate()
-                .map(|(id, entry)| {
-                    match entry {
-                        EvaluationOutput::Error(err) => {
-                            format!("[{}] Error: {}\n", id, err.root_cause())
-                        }
-                        _ => format!("[{}] {}\n", id, entry),
-                    }
-                })
-                .filter(|x| print_stage(&x))
-                .collect::<String>()
-                .lines()
-                .map(|line| format!("    {}\n", line))
-                .collect::<String>()
-        )?;
-    } else {
-        for (id, entry) in res.text.iter().enumerate() {
-            if print_stage(entry) {
-                writeln!(output, "    [{}] {}", id, entry)?;
+    let mut exp_file_path = PathBuf::from(path);
+    exp_file_path.set_extension("exp");
+
+    if exp_file_path.exists() {
+        // REVIEW: "\n" is probably not very portable.
```

Comment:
How hard would it be to split this up into separate functions (e.g., one for handling .exp files and the other for embedded checks)? I think that might help make this more readable and maintainable, if it isn't too hard to split up that way.

Reply:
Actually, that same comment applies later in this function, too. It's just such a long sequence of code that it's hard to quickly see the overall logic.

Reply:
It shouldn't be hard -- a few macros needs to capture some bindings but it's not that troublesome. Let me try.